File: backend/models/Center.js
import mongoose from 'mongoose';

const centerScheduleSchema = new mongoose.Schema({
  workingDays: {
    type: [String],
    default: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday']
  },
  startTime: {
    type: String,
    default: '08:00'
  },
  endTime: {
    type: String,
    default: '16:00'
  },
  breakTime: {
    start: {
      type: String,
      default: ''
    },
    end: {
      type: String,
      default: ''
    }
  },
  slotDuration: {
    type: Number,
    default: 30,
    min: 5
  },
  maxBookingsPerSlot: {
    type: Number,
    default: 5,
    min: 1
  },
  maxAppointmentsPerDay: {
    type: Number,
    default: 100,
    min: 1
  },
  closedDays: {
    type: [String],
    default: ['Friday']
  },
  closedDates: {
    type: [String],
    default: []
  },
  specialUnavailableDates: {
    type: [String],
    default: []
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, { _id: false });

const centerSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Please add a center name'],
    unique: true,
    trim: true
  },
  address: {
    type: String,
    required: [true, 'Please add an address']
  },
  city: {
    type: String,
    required: [true, 'Please add a city']
  },
  district: {
    type: String,
    required: [true, 'Please select a district'],
    trim: true
  },
  phone: {
    type: String,
    required: [true, 'Please add a phone number']
  },
  counters: {
    type: Number,
    required: [true, 'Please add number of active counters'],
    default: 5
  },
  capacity: {
    type: Number,
    required: [true, 'Please add daily capacity limit'],
    default: 100
  },
  hours: {
    type: String,
    required: [true, 'Please add operating hours'],
    default: '08:00 AM - 05:00 PM'
  },
  status: {
    type: String,
    enum: ['Active', 'Inactive', 'Maintenance', 'Closed'],
    default: 'Active'
  },
  schedule: {
    type: centerScheduleSchema,
    default: () => ({})
  }
}, {
  timestamps: true
});

const Center = mongoose.model('Center', centerSchema);
export default Center;

File: mobile/lib/shared/models/center.dart
import '../../core/utils/json_utils.dart';

/// `GET /api/centers` item, produced by `serializeCenter` on the backend.
class CenterModel {
  const CenterModel({
    required this.id,
    required this.name,
    this.address = '',
    this.city = '',
    this.district = '',
    this.phone = '',
    this.counters = 0,
    this.capacity = 0,
    this.hours = '',
    this.status = 'Active',
    this.schedule = const CenterSchedule(),
    this.currentQueueSize = 0,
    this.estimatedWaitMinutes = 0,
    this.availableServices = const [],
  });

  final String id;
  final String name;
  final String address;
  final String city;
  final String district;
  final String phone;
  final int counters;
  final int capacity;
  final String hours;
  final String status;
  final CenterSchedule schedule;
  final int currentQueueSize;
  final int estimatedWaitMinutes;
  final List<String> availableServices;

  bool get isActive => status.toLowerCase() == 'active' && schedule.isActive;

  String get location {
    final parts = [address, city, district].where((p) => p.trim().isNotEmpty).toSet();
    return parts.isEmpty ? 'Location not provided' : parts.join(', ');
  }

  /// Citizen-facing Open / Busy / Closed for the centers list.
  String get operationalStatus {
    if (!isActive || !schedule.isOpenOn(DateTime.now())) return 'Closed';
    final queue = displayQueueSize;
    final busyThreshold = capacity > 0
        ? (capacity * 0.7).ceil()
        : (schedule.maxAppointmentsPerDay * 0.7).ceil();
    if (queue > 0 && queue >= busyThreshold) return 'Busy';
    return 'Open';
  }

  int get displayQueueSize {
    if (currentQueueSize > 0) return currentQueueSize;
    return 0;
  }

  String get displayEstimatedWait {
    if (estimatedWaitMinutes > 0) return '$estimatedWaitMinutes min';
    final queue = displayQueueSize;
    if (queue <= 0) return '~${schedule.slotDuration} min';
    return '~${queue * schedule.slotDuration} min';
  }

  List<String> get displayServices {
    if (availableServices.isNotEmpty) return availableServices;
    return const [
      'New Registration',
      'Update Information',
      'Lost ID Replacement',
    ];
  }

  String get hoursLabel {
    if (hours.trim().isNotEmpty) return hours;
    return '${schedule.startTime} – ${schedule.endTime}';
  }

  factory CenterModel.fromJson(Map<String, dynamic> json) {
    final servicesRaw = json['availableServices'] ?? json['services'];
    return CenterModel(
      id: Json.idOf(json),
      name: Json.str(json['name']),
      address: Json.str(json['address']),
      city: Json.str(json['city']),
      district: Json.str(json['district']),
      phone: Json.str(json['phone']),
      counters: Json.intOf(json['counters']),
      capacity: Json.intOf(json['capacity']),
      hours: Json.str(json['hours']),
      status: Json.str(json['status'], 'Active'),
      schedule: CenterSchedule.fromJson(Json.map(json['schedule'])),
      currentQueueSize: Json.intOf(
        json['currentQueueSize'] ?? json['queueSize'] ?? json['waitingCount'],
      ),
      estimatedWaitMinutes: Json.intOf(
        json['estimatedWaitMinutes'] ?? json['estimatedWait'],
      ),
      availableServices: servicesRaw is List
          ? servicesRaw.map((e) => '$e').where((e) => e.trim().isNotEmpty).toList()
          : const [],
    );
  }
}

class CenterSchedule {
  const CenterSchedule({
    this.workingDays = const ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday'],
    this.startTime = '08:00',
    this.endTime = '16:00',
    this.slotDuration = 30,
    this.maxBookingsPerSlot = 5,
    this.maxAppointmentsPerDay = 100,
    this.closedDays = const ['Friday', 'Saturday'],
    this.closedDates = const [],
    this.isActive = true,
  });

  final List<String> workingDays;
  final String startTime;
  final String endTime;
  final int slotDuration;
  final int maxBookingsPerSlot;
  final int maxAppointmentsPerDay;
  final List<String> closedDays;
  final List<String> closedDates;
  final bool isActive;

  static const _weekdays = [
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday',
    'Sunday',
  ];

  /// True when this center accepts appointments on [date]
  /// (working day and not in closed/holiday dates).
  bool isOpenOn(DateTime date) {
    if (!isActive) return false;
    final day = DateTime(date.year, date.month, date.day);
    final weekday = _weekdays[day.weekday - 1];
    final key =
        '${day.year.toString().padLeft(4, '0')}-${day.month.toString().padLeft(2, '0')}-${day.day.toString().padLeft(2, '0')}';
    final closed = {
      for (final raw in closedDates)
        if (raw.trim().length >= 10) raw.trim().substring(0, 10),
    };
    if (closed.contains(key)) return false;
    if (closedDays.contains(weekday)) return false;
    if (workingDays.isNotEmpty && !workingDays.contains(weekday)) return false;
    return true;
  }

  factory CenterSchedule.fromJson(Map<String, dynamic> json) {
    if (json.isEmpty) return const CenterSchedule();
    final working = Json.stringList(json['workingDays']);
    final closedDayList = Json.stringList(json['closedDays']);
    return CenterSchedule(
      workingDays: working.isEmpty
          ? const ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday']
          : working,
      startTime: Json.str(json['startTime'], '08:00'),
      endTime: Json.str(json['endTime'], '16:00'),
      slotDuration: Json.intOf(json['slotDuration'], 30),
      maxBookingsPerSlot: Json.intOf(json['maxBookingsPerSlot'], 5),
      maxAppointmentsPerDay: Json.intOf(json['maxAppointmentsPerDay'], 100),
      closedDays: closedDayList,
      closedDates: Json.stringList(json['closedDates'])
          .map((d) => d.trim())
          .where((d) => d.length >= 10)
          .map((d) => d.substring(0, 10))
          .toList(),
      isActive: Json.boolOf(json['isActive'], fallback: true),
    );
  }
}
